File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/groupsv2/GroupChangeActionsBuilderChangeSetModifier.java
package org.whispersystems.signalservice.api.groupsv2;

import org.signal.storageservice.protos.groups.GroupChange;

final class GroupChangeActionsBuilderChangeSetModifier implements ChangeSetModifier {
  private final GroupChange.Actions.Builder result;

  public GroupChangeActionsBuilderChangeSetModifier(GroupChange.Actions.Builder result) {
    this.result = result;
  }

  @Override
  public void removeAddMembers(int i) {
    result.removeAddMembers(i);
  }

  @Override
  public void moveAddToPromote(int i) {
    GroupChange.Actions.AddMemberAction addMemberAction = result.getAddMembersList().get(i);
    result.removeAddMembers(i);
    result.addPromotePendingMembers(GroupChange.Actions.PromotePendingMemberAction.newBuilder().setPresentation(addMemberAction.getAdded().getPresentation()));
  }

  @Override
  public void removeDeleteMembers(int i) {
    result.removeDeleteMembers(i);
  }

  @Override
  public void removeModifyMemberRoles(int i) {
    result.removeModifyMemberRoles(i);
  }

  @Override
  public void removeModifyMemberProfileKeys(int i) {
    result.removeModifyMemberProfileKeys(i);
  }

  @Override
  public void removeAddPendingMembers(int i) {
    result.removeAddPendingMembers(i);
  }

  @Override
  public void removeDeletePendingMembers(int i) {
    result.removeDeletePendingMembers(i);
  }

  @Override
  public void removePromotePendingMembers(int i) {
    result.removePromotePendingMembers(i);
  }

  @Override
  public void clearModifyTitle() {
    result.clearModifyTitle();
  }

  @Override
  public void clearModifyAvatar() {
    result.clearModifyAvatar();
  }

  @Override
  public void clearModifyDisappearingMessagesTimer() {
    result.clearModifyDisappearingMessagesTimer();
  }

  @Override
  public void clearModifyAttributesAccess() {
    result.clearModifyAttributesAccess();
  }

  @Override
  public void clearModifyMemberAccess() {
    result.clearModifyMemberAccess();
  }

  @Override
  public void clearModifyAddFromInviteLinkAccess() {
    result.clearModifyAddFromInviteLinkAccess();
  }

  @Override
  public void removeAddRequestingMembers(int i) {
    result.removeAddRequestingMembers(i);
  }

  @Override
  public void moveAddRequestingMembersToPromote(int i) {
    GroupChange.Actions.AddRequestingMemberAction addMemberAction = result.getAddRequestingMembersList().get(i);
    result.removeAddRequestingMembers(i);
    result.addPromotePendingMembers(0, GroupChange.Actions.PromotePendingMemberAction.newBuilder().setPresentation(addMemberAction.getAdded().getPresentation()));
  }

  @Override
  public void removeDeleteRequestingMembers(int i) {
    result.removeDeleteRequestingMembers(i);
  }

  @Override
  public void removePromoteRequestingMembers(int i) {
    result.removePromoteRequestingMembers(i);
  }
}

File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/groupsv2/GroupChangeActionsBuilderChangeSetModifier.kt
package org.whispersystems.signalservice.api.groupsv2

import org.signal.storageservice.protos.groups.GroupChange
import org.signal.storageservice.protos.groups.GroupChange.Actions.AddMemberAction
import org.signal.storageservice.protos.groups.GroupChange.Actions.AddRequestingMemberAction

internal class GroupChangeActionsBuilderChangeSetModifier(private val result: GroupChange.Actions.Builder) : ChangeSetModifier {
  override fun removeAddMembers(i: Int) {
    result.addMembers = result.addMembers.removeIndex(i)
  }

  override fun moveAddToPromote(i: Int) {
    val addMemberAction: AddMemberAction = result.addMembers[i]
    result.addMembers = result.addMembers.removeIndex(i)
    result.promotePendingMembers += GroupChange.Actions.PromotePendingMemberAction.Builder().presentation(addMemberAction.added!!.presentation).build()
  }

  override fun removeDeleteMembers(i: Int) {
    result.deleteMembers = result.deleteMembers.removeIndex(i)
  }

  override fun removeModifyMemberRoles(i: Int) {
    result.modifyMemberRoles = result.modifyMemberRoles.removeIndex(i)
  }

  override fun removeModifyMemberProfileKeys(i: Int) {
    result.modifyMemberProfileKeys = result.modifyMemberProfileKeys.removeIndex(i)
  }

  override fun removeAddPendingMembers(i: Int) {
    result.addPendingMembers = result.addPendingMembers.removeIndex(i)
  }

  override fun removeDeletePendingMembers(i: Int) {
    result.deletePendingMembers = result.deletePendingMembers.removeIndex(i)
  }

  override fun removePromotePendingMembers(i: Int) {
    result.promotePendingMembers = result.promotePendingMembers.removeIndex(i)
  }

  override fun clearModifyTitle() {
    result.modifyTitle = null
  }

  override fun clearModifyAvatar() {
    result.modifyAvatar = null
  }

  override fun clearModifyDisappearingMessagesTimer() {
    result.modifyDisappearingMessagesTimer = null
  }

  override fun clearModifyAttributesAccess() {
    result.modifyAttributesAccess = null
  }

  override fun clearModifyMemberAccess() {
    result.modifyMemberAccess = null
  }

  override fun clearModifyAddFromInviteLinkAccess() {
    result.modifyAddFromInviteLinkAccess = null
  }

  override fun removeAddRequestingMembers(i: Int) {
    result.addRequestingMembers = result.addRequestingMembers.removeIndex(i)
  }

  override fun moveAddRequestingMembersToPromote(i: Int) {
    val addMemberAction: AddRequestingMemberAction = result.addRequestingMembers[i]
    result.addRequestingMembers = result.addRequestingMembers.removeIndex(i)
    result.promotePendingMembers += GroupChange.Actions.PromotePendingMemberAction.Builder().presentation(addMemberAction.added!!.presentation).build()
  }

  override fun removeDeleteRequestingMembers(i: Int) {
    result.deleteRequestingMembers = result.deleteRequestingMembers.removeIndex(i)
  }

  override fun removePromoteRequestingMembers(i: Int) {
    result.promoteRequestingMembers = result.promoteRequestingMembers.removeIndex(i)
  }

  override fun clearModifyDescription() {
    result.modifyDescription = null
  }

  override fun clearModifyAnnouncementsOnly() {
    result.modifyAnnouncementsOnly = null
  }

  override fun removeAddBannedMembers(i: Int) {
    result.addBannedMembers = result.addBannedMembers.removeIndex(i)
  }

  override fun removeDeleteBannedMembers(i: Int) {
    result.deleteBannedMembers = result.deleteBannedMembers.removeIndex(i)
  }

  override fun removePromotePendingPniAciMembers(i: Int) {
    result.promotePendingPniAciMembers = result.promotePendingPniAciMembers.removeIndex(i)
  }

  private fun <T> List<T>.removeIndex(i: Int): List<T> {
    val modifiedList = this.toMutableList()
    modifiedList.removeAt(i)
    return modifiedList
  }
}
